"use strict";(self.webpackChunkStatescope=self.webpackChunkStatescope||[]).push([[2782],{7997:(e,n,t)=>{t.r(n),t.d(n,{assets:()=>o,contentTitle:()=>l,default:()=>p,frontMatter:()=>a,metadata:()=>s,toc:()=>c});const s=JSON.parse('{"id":"input","title":"Input Arguments","description":"This page outlines the required inputs and optional parameters for running Statescope.","source":"@site/docs/input.md","sourceDirName":".","slug":"/input","permalink":"/Statescopeweb/docs/input","draft":false,"unlisted":false,"editUrl":"https://github.com/tgac-vumc/Statescopeweb.git/docs/input.md","tags":[],"version":"current","frontMatter":{"id":"input","title":"Input Arguments","sidebar_label":"Input Arguments"},"sidebar":"tutorialSidebar","previous":{"title":"Installation","permalink":"/Statescopeweb/docs/installation"},"next":{"title":"Python","permalink":"/Statescopeweb/docs/python"}}');var r=t(4848),i=t(8453);const a={id:"input",title:"Input Arguments",sidebar_label:"Input Arguments"},l="Input Arguments",o={},c=[{value:"Required Datasets",id:"required-datasets",level:2},{value:"1. Signature Matrices",id:"1-signature-matrices",level:3},{value:"2. Bulk Gene Expression Data",id:"2-bulk-gene-expression-data",level:3},{value:"3. Expected Cell Fractions (Optional)",id:"3-expected-cell-fractions-optional",level:3},{value:"Options for Reference RNA-seq Data",id:"options-for-reference-rna-seq-data",level:2}];function d(e){const n={a:"a",em:"em",h1:"h1",h2:"h2",h3:"h3",header:"header",li:"li",ol:"ol",p:"p",strong:"strong",ul:"ul",...(0,i.R)(),...e.components};return(0,r.jsxs)(r.Fragment,{children:[(0,r.jsx)(n.header,{children:(0,r.jsx)(n.h1,{id:"input-arguments",children:"Input Arguments"})}),"\n",(0,r.jsxs)(n.p,{children:["This page outlines the required inputs and optional parameters for running ",(0,r.jsx)(n.strong,{children:"Statescope"}),"."]}),"\n",(0,r.jsx)(n.h2,{id:"required-datasets",children:"Required Datasets"}),"\n",(0,r.jsx)(n.h3,{id:"1-signature-matrices",children:"1. Signature Matrices"}),"\n",(0,r.jsxs)(n.ul,{children:["\n",(0,r.jsxs)(n.li,{children:["\n",(0,r.jsxs)(n.p,{children:[(0,r.jsx)(n.strong,{children:"X"}),": An ",(0,r.jsx)(n.em,{children:"Ngene"})," by ",(0,r.jsx)(n.em,{children:"Ncell"})," matrix containing average gene expression profiles per cell type (a signature matrix) in ",(0,r.jsx)(n.strong,{children:"log-scale"}),"."]}),"\n"]}),"\n",(0,r.jsxs)(n.li,{children:["\n",(0,r.jsxs)(n.p,{children:[(0,r.jsx)(n.strong,{children:"stdX"}),": An ",(0,r.jsx)(n.em,{children:"Ngene"})," by ",(0,r.jsx)(n.em,{children:"Ncell"})," matrix containing standard deviation per gene per cell type (a signature matrix of gene expression variability)."]}),"\n"]}),"\n"]}),"\n",(0,r.jsx)(n.h3,{id:"2-bulk-gene-expression-data",children:"2. Bulk Gene Expression Data"}),"\n",(0,r.jsxs)(n.ul,{children:["\n",(0,r.jsxs)(n.li,{children:[(0,r.jsx)(n.strong,{children:"Y"}),": An ",(0,r.jsx)(n.em,{children:"Ngene"})," by ",(0,r.jsx)(n.em,{children:"Nsample"})," matrix containing bulk gene expression data. This should be in ",(0,r.jsx)(n.strong,{children:"linear-scale"})," data without log-transformation."]}),"\n"]}),"\n",(0,r.jsx)(n.h3,{id:"3-expected-cell-fractions-optional",children:"3. Expected Cell Fractions (Optional)"}),"\n",(0,r.jsxs)(n.ul,{children:["\n",(0,r.jsxs)(n.li,{children:[(0,r.jsx)(n.strong,{children:"Expectation"}),": An ",(0,r.jsx)(n.em,{children:"Nsample"})," by ",(0,r.jsx)(n.em,{children:"Ncell"})," matrix containing the expected cell fractions used to inform ",(0,r.jsx)(n.strong,{children:"OncoBLADE"})," (optional)."]}),"\n"]}),"\n",(0,r.jsx)(n.h2,{id:"options-for-reference-rna-seq-data",children:"Options for Reference RNA-seq Data"}),"\n",(0,r.jsx)(n.p,{children:"Users have two options for obtaining signature matrices:"}),"\n",(0,r.jsxs)(n.ol,{children:["\n",(0,r.jsxs)(n.li,{children:["\n",(0,r.jsx)(n.p,{children:(0,r.jsx)(n.strong,{children:"Use Pre-processed Signatures"})}),"\n",(0,r.jsxs)(n.ul,{children:["\n",(0,r.jsxs)(n.li,{children:["Proceed to the ",(0,r.jsx)(n.a,{href:"/Statescopeweb/docs/processed-signatures",children:"Processed Dataset"})," page to download and use pre-processed signature matrices."]}),"\n"]}),"\n"]}),"\n",(0,r.jsxs)(n.li,{children:["\n",(0,r.jsx)(n.p,{children:(0,r.jsx)(n.strong,{children:"Create Signatures from scRNA Data"})}),"\n",(0,r.jsxs)(n.ul,{children:["\n",(0,r.jsxs)(n.li,{children:["Follow the instructions on the ",(0,r.jsx)(n.a,{href:"/Statescopeweb/docs/python",children:"Python Tutorial"})," page to generate signatures from single-cell RNA-seq data."]}),"\n"]}),"\n"]}),"\n"]})]})}function p(e={}){const{wrapper:n}={...(0,i.R)(),...e.components};return n?(0,r.jsx)(n,{...e,children:(0,r.jsx)(d,{...e})}):d(e)}},8453:(e,n,t)=>{t.d(n,{R:()=>a,x:()=>l});var s=t(6540);const r={},i=s.createContext(r);function a(e){const n=s.useContext(i);return s.useMemo((function(){return"function"==typeof e?e(n):{...n,...e}}),[n,e])}function l(e){let n;return n=e.disableParentContext?"function"==typeof e.components?e.components(r):e.components||r:a(e.components),s.createElement(i.Provider,{value:n},e.children)}}}]);