"use strict";(self.webpackChunkStatescope=self.webpackChunkStatescope||[]).push([[8907],{9858:e=>{e.exports=JSON.parse('{"version":{"pluginId":"default","version":"current","label":"Next","banner":null,"badge":false,"noIndex":false,"className":"docs-version-current","isLast":true,"docsSidebars":{"tutorialSidebar":[{"type":"link","label":"Installation","href":"/Statescopeweb/docs/installation","docId":"installation","unlisted":false},{"type":"link","label":"Input Arguments","href":"/Statescopeweb/docs/input","docId":"input","unlisted":false},{"type":"category","label":"Tutorials","items":[{"type":"category","label":"Tutorials Simple","items":[{"type":"link","label":"Python","href":"/Statescopeweb/docs/python","docId":"python","unlisted":false},{"type":"link","label":"R","href":"/Statescopeweb/docs/r","docId":"r","unlisted":false}],"collapsed":true,"collapsible":true},{"type":"category","label":"Tutorials Advanced Users ","items":[{"type":"link","label":"Create Signatures","href":"/Statescopeweb/docs/create-signatures","docId":"create-signatures","unlisted":false},{"type":"link","label":"Run Deconvolution","href":"/Statescopeweb/docs/deconvolution","docId":"deconvolution","unlisted":false},{"type":"link","label":"Run Purification","href":"/Statescopeweb/docs/purification","docId":"purification","unlisted":false}],"collapsed":true,"collapsible":true}],"collapsed":true,"collapsible":true},{"type":"link","label":"Troubleshooting","href":"/Statescopeweb/docs/troubleshooting","docId":"troubleshooting","unlisted":false}]},"docs":{"about-us":{"id":"about-us","title":"About Us","description":"Research Focus"},"create-signatures":{"id":"create-signatures","title":"Create Signatures from scRNA Data","description":"This page guides you through the process of creating signature matrices (X and stdX) from your own single-cell RNA-seq data. Additionally, instructions for using a Snakemake pipeline and optionally scheduling it with SLURM on the HPC are provided.","sidebar":"tutorialSidebar"},"datasets":{"id":"datasets","title":"Datasets","description":"In-House Processed scRNA Expression Profiles"},"deconvolution":{"id":"deconvolution","title":"Deconvolution","description":"This page provides an overview of the deconvolution process in BLADE and instructions on how to apply it to your data using the provided Snakemake pipeline.","sidebar":"tutorialSidebar"},"input":{"id":"input","title":"Input Arguments","description":"This page outlines the required inputs and optional parameters for running OncoBLADE.","sidebar":"tutorialSidebar"},"installation":{"id":"installation","title":"Installation Guide","description":"Hardware Requirements","sidebar":"tutorialSidebar"},"intro":{"id":"intro","title":"Tutorial Intro","description":"Introduction"},"overview":{"id":"overview","title":"Overview","description":"In the BLADE package, you can load the following functions and modules."},"PBMC-deconvolution":{"id":"PBMC-deconvolution","title":"Deconvolution of PBMC Data","description":""},"processed-signatures":{"id":"processed-signatures","title":"Processed Signatures","description":"This page hosts our in-house single-cell RNA (scRNA) signatures for various applications. Below, you will find details about the processed signatures, including their source, cell type information, gene count, and download links."},"purification":{"id":"purification","title":"purification","description":"","sidebar":"tutorialSidebar"},"python":{"id":"python","title":"Python tutorial for Statescope","description":"In this tutorial, we will walk through setting up and running the Statescope package in Python for bulk data analysis. We will cover installation, environment setup, data import, deconvolution, refinement, and cell state discovery.","sidebar":"tutorialSidebar"},"r":{"id":"r","title":"r","description":"","sidebar":"tutorialSidebar"},"troubleshooting":{"id":"troubleshooting","title":"troubleshooting","description":"","sidebar":"tutorialSidebar"},"tutorial-basics/congratulations":{"id":"tutorial-basics/congratulations","title":"Congratulations!","description":"You have just learned the basics of Docusaurus and made some changes to the initial template."},"tutorial-basics/create-a-blog-post":{"id":"tutorial-basics/create-a-blog-post","title":"Create a Blog Post","description":"Docusaurus creates a page for each blog post, but also a blog index page, a tag system, an RSS feed..."},"tutorial-basics/create-a-document":{"id":"tutorial-basics/create-a-document","title":"Create a Document","description":"Documents are groups of pages connected through:"},"tutorial-basics/create-a-page":{"id":"tutorial-basics/create-a-page","title":"Create a Page","description":"Add Markdown or React files to src/pages to create a standalone page:"},"tutorial-basics/deploy-your-site":{"id":"tutorial-basics/deploy-your-site","title":"Deploy your site","description":"Docusaurus is a static-site-generator (also called Jamstack)."},"tutorial-basics/markdown-features":{"id":"tutorial-basics/markdown-features","title":"Markdown Features","description":"Docusaurus supports Markdown and a few additional features."},"tutorial-extras/manage-docs-versions":{"id":"tutorial-extras/manage-docs-versions","title":"Manage Docs Versions","description":"Docusaurus can manage multiple versions of your docs."},"tutorial-extras/translate-your-site":{"id":"tutorial-extras/translate-your-site","title":"Translate your site","description":"Let\'s translate docs/intro.md to French."}}}}')}}]);